### ISSUE TITLE
ConversionException in network, when sending to another agent an agent with a non-empty mailbox

### ISSUE THREAD

**Issue by benoitgaudou** [gama-platform/gama#2828]
**Describe the bug**

**To Reproduce**
1. Run the model `Example_MQTT_Send.gaml`  and `Example_MQTT_Receive.gaml` from the network plugin
2. Run both of them
3. The sender sends first itself, then a string and then itself again.
When the receiver receives the agent the second time, it throws the following exception.

```
com.thoughtworks.xstream.converters.ConversionException: Invalid reference
---- Debugging information ----
reference           : ../../../../../../../../../..
class               : msi.gama.extensions.messaging.GamaMessage
required-type       : msi.gama.extensions.messaging.GamaMessage
converter-type      : com.thoughtworks.xstream.converters.reflection.ReflectionConverter
path                : /msi.gama.metamodel.agent.GamlAgent/variables/data/msi.gama.metamodel.shape.GamaShape/attributes/valuesMapReducer/entry/msi.gama.extensions.messaging.GamaMailbox/valuesListReducer/msi.gama.extensions.messaging.GamaMessage/sender
class[1]            : java.util.ArrayList
converter-type[1]   : com.thoughtworks.xstream.converters.collections.CollectionConverter
class[2]            : ummisco.gama.serializer.gamaType.reduced.GamaListReducerNetwork
class[3]            : msi.gama.extensions.messaging.GamaMailbox
converter-type[2]   : ummisco.gama.serializer.gamaType.converters.GamaListConverterNetwork
class[4]            : java.util.HashMap
converter-type[3]   : com.thoughtworks.xstream.converters.collections.MapConverter
class[5]            : ummisco.gama.serializer.gamaType.reduced.GamaMapReducer
class[6]            : msi.gama.util.GamaMap
converter-type[4]   : ummisco.gama.serializer.gamaType.converters.GamaMapConverter
class[7]            : msi.gama.metamodel.shape.GamaShape
class[8]            : msi.gama.metamodel.agent.SavedAgent
converter-type[5]   : ummisco.gama.serializer.gamaType.converters.SavedAgentConverter
class[9]            : msi.gama.metamodel.agent.GamlAgent
converter-type[6]   : ummisco.gama.serializer.gamaType.converters.GamaAgentConverterNetwork
version             : 1.4.8
-------------------------------
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:57)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:71)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.addCurrentElementToCollection(CollectionConverter.java:98)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:91)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:85)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.unmarshal(CollectionConverter.java:80)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at ummisco.gama.serializer.gamaType.converters.GamaListConverterNetwork.unmarshal(GamaListConverterNetwork.java:61)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:71)
	at com.thoughtworks.xstream.converters.collections.MapConverter.putCurrentEntryIntoMap(MapConverter.java:110)
	at com.thoughtworks.xstream.converters.collections.MapConverter.populateMap(MapConverter.java:98)
	at com.thoughtworks.xstream.converters.collections.MapConverter.populateMap(MapConverter.java:92)
	at com.thoughtworks.xstream.converters.collections.MapConverter.unmarshal(MapConverter.java:87)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at ummisco.gama.serializer.gamaType.converters.GamaMapConverter.unmarshal(GamaMapConverter.java:64)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:71)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.addCurrentElementToCollection(CollectionConverter.java:98)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:91)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:85)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.unmarshal(CollectionConverter.java:80)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at ummisco.gama.serializer.gamaType.converters.SavedAgentConverter.unmarshal(SavedAgentConverter.java:85)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at ummisco.gama.serializer.gamaType.converters.GamaAgentConverterNetwork.unmarshal(GamaAgentConverterNetwork.java:58)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.start(TreeUnmarshaller.java:134)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.unmarshal(AbstractTreeMarshallingStrategy.java:32)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1206)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1190)
	at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1061)
	at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1052)
	at ummisco.gama.serializer.factory.StreamConverter.convertNetworkStreamToObject(StreamConverter.java:79)
	at ummisco.gama.network.common.CompositeGamaMessage.getContents(CompositeGamaMessage.java:39)
	at gaml.additions.GamlAdditions.lambda$232(GamlAdditions.java:304)
	at msi.gaml.expressions.UnaryOperator._value(UnaryOperator.java:83)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.expressions.BinaryOperator._value(BinaryOperator.java:76)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.statements.WriteStatement.privateExecuteIn(WriteStatement.java:88)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:41)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.runtime.IScope.execute(IScope.java:506)
	at msi.gaml.statements.AbstractStatementSequence.privateExecuteIn(AbstractStatementSequence.java:56)
	at msi.gaml.architecture.reflex.ReflexStatement.privateExecuteIn(ReflexStatement.java:81)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:41)
	at msi.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:46)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.runtime.IScope.execute(IScope.java:506)
	at msi.gaml.architecture.reflex.ReflexArchitecture.executeReflexes(ReflexArchitecture.java:106)
	at msi.gaml.architecture.reflex.ReflexArchitecture.executeOn(ReflexArchitecture.java:99)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.metamodel.agent.AbstractAgent.doStep(AbstractAgent.java:283)
	at msi.gama.metamodel.agent.MinimalAgent.doStep(MinimalAgent.java:235)
	at msi.gama.metamodel.agent.AbstractAgent.step(AbstractAgent.java:253)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:451)
	at msi.gama.runtime.concurrent.GamaExecutorService.doStep(GamaExecutorService.java:236)
	at msi.gama.runtime.concurrent.GamaExecutorService.step(GamaExecutorService.java:204)
	at msi.gama.metamodel.population.GamaPopulation.stepAgents(GamaPopulation.java:208)
	at msi.gama.metamodel.population.GamaPopulation.step(GamaPopulation.java:203)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:415)
	at msi.gama.metamodel.agent.GamlAgent.stepSubPopulations(GamlAgent.java:98)
	at msi.gama.metamodel.agent.AbstractAgent.doStep(AbstractAgent.java:283)
	at msi.gama.metamodel.agent.MinimalAgent.doStep(MinimalAgent.java:235)
	at msi.gama.metamodel.agent.AbstractAgent.step(AbstractAgent.java:253)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:451)
	at msi.gama.runtime.concurrent.SimulationRunner.lambda$0(SimulationRunner.java:30)
	at msi.gama.runtime.concurrent.SimulationRunner.lambda$1(SimulationRunner.java:55)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Java error: ConversionException
	at msi.gama.runtime.exceptions.GamaRuntimeException.create(GamaRuntimeException.java:58)
	at msi.gaml.expressions.UnaryOperator._value(UnaryOperator.java:90)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.expressions.BinaryOperator._value(BinaryOperator.java:76)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.statements.WriteStatement.privateExecuteIn(WriteStatement.java:88)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:41)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.runtime.IScope.execute(IScope.java:506)
	at msi.gaml.statements.AbstractStatementSequence.privateExecuteIn(AbstractStatementSequence.java:56)
	at msi.gaml.architecture.reflex.ReflexStatement.privateExecuteIn(ReflexStatement.java:81)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:41)
	at msi.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:46)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.runtime.IScope.execute(IScope.java:506)
	at msi.gaml.architecture.reflex.ReflexArchitecture.executeReflexes(ReflexArchitecture.java:106)
	at msi.gaml.architecture.reflex.ReflexArchitecture.executeOn(ReflexArchitecture.java:99)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:369)
	at msi.gama.metamodel.agent.AbstractAgent.doStep(AbstractAgent.java:283)
	at msi.gama.metamodel.agent.MinimalAgent.doStep(MinimalAgent.java:235)
	at msi.gama.metamodel.agent.AbstractAgent.step(AbstractAgent.java:253)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:451)
	at msi.gama.runtime.concurrent.GamaExecutorService.doStep(GamaExecutorService.java:236)
	at msi.gama.runtime.concurrent.GamaExecutorService.step(GamaExecutorService.java:204)
	at msi.gama.metamodel.population.GamaPopulation.stepAgents(GamaPopulation.java:208)
	at msi.gama.metamodel.population.GamaPopulation.step(GamaPopulation.java:203)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:415)
	at msi.gama.metamodel.agent.GamlAgent.stepSubPopulations(GamlAgent.java:98)
	at msi.gama.metamodel.agent.AbstractAgent.doStep(AbstractAgent.java:283)
	at msi.gama.metamodel.agent.MinimalAgent.doStep(MinimalAgent.java:235)
	at msi.gama.metamodel.agent.AbstractAgent.step(AbstractAgent.java:253)
	at msi.gama.runtime.ExecutionScope.step(ExecutionScope.java:451)
	at msi.gama.runtime.concurrent.SimulationRunner.lambda$0(SimulationRunner.java:30)
	at msi.gama.runtime.concurrent.SimulationRunner.lambda$1(SimulationRunner.java:55)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Desktop (please complete the following information):**
 - OS: macOS
 - GAMA version: git

**Comment by gnoubi** [comment]
Was something been modified in serialiser ?

Nicolas 

Nicolas Marilleau
UMI 209 UMMISCO IRD/UPMC
Chercheur invité Femto-st

Institut de Recherche pour le Développement 
32 avenue Henri Varagnat
93143 Bondy 
Fixe : 0148027901
Mobile: 0688334906

**Comment by benoitgaudou** [comment]
After deeper investigation, it seems that that the issue occurs when:
- an agent with the network skill is sent to an agent in another simulation
- that its mailbox is not empty

**Comment by AlexisDrogoul** [comment]
I checked and older versions (I tested that of the 22nd of June) also show the same error. @hqnghi88, if you have even older ones at hand, can you test ?

**Comment by hqnghi88** [comment]
I test with RC2 , open 2 gama on 2 workspaces, launch Receive, launch Send. After some step of Receiver, i have this exception in GAMA. Sorry i dont have the eclipse of rc2 . AND i dont have mqtt server on my machine.
2 occurrences in 2 agents at cycle 1: Java error: ConversionException
in write (name + ' fecth this message: ') + mess.contents ;
ConversionException: Cannot deserialize object with new readObject()/writeObject() methods
---- Debugging information ----
class               : msi.gama.extensions.messaging.GamaMailbox
required-type       : msi.gama.extensions.messaging.GamaMailbox
converter-type      : com.thoughtworks.xstream.converters.reflection.SerializableConverter
path                : /msi.gama.metamodel.agent.MinimalAgent/variables/data/msi.gama.metamodel.shape.GamaShape/attributes/valuesMapReducer/entry/msi.gama.extensions.messaging.GamaMailbox
class[1]            : java.util.HashMap
converter-type[1]   : com.thoughtworks.xstream.converters.collections.MapConverter
class[2]            : ummisco.gama.serializer.gamaType.reduced.GamaMapReducer
converter-type[2]   : com.thoughtworks.xstream.converters.reflection.ReflectionConverter
class[3]            : msi.gama.util.GamaMap
converter-type[3]   : ummisco.gama.serializer.gamaType.converters.GamaMapConverter
class[4]            : msi.gama.metamodel.shape.GamaShape
class[5]            : java.util.ArrayList
converter-type[4]   : com.thoughtworks.xstream.converters.collections.CollectionConverter
class[6]            : msi.gama.metamodel.agent.SavedAgent
converter-type[5]   : ummisco.gama.serializer.gamaType.converters.SavedAgentConverter
class[7]            : msi.gama.metamodel.agent.MinimalAgent
converter-type[6]   : ummisco.gama.serializer.gamaType.converters.GamaAgentConverterNetwork
version             : 1.4.8
-------------------------------
com.thoughtworks.xstream.converters.reflection.SerializableConverter.doUnmarshal(SerializableConverter.java:330)
com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:71)
when applying the contents operator on ummisco.gama.network.common.CompositeGamaMessage@51bf91de
in write (name + ' fecth this message: ') + mess.contents ;
in reflex fetch when: self.has_more_message()  {
	message mess <- self.fetch_message();
	write (name + ' fecth this message: ') + mess.contents ;
}

in agents NetworkingAgent0, Simulation 0

**Comment by gnoubi** [comment]
Could be due to the serialisation of gamaMessage that can contain iAgent as sender and receiver ?

I think it is the reason of “referenced agent” in message.


Nicolas Marilleau
UMI 209 UMMISCO IRD/UPMC
Chercheur invité Femto-st

Institut de Recherche pour le Développement 
32 avenue Henri Varagnat
93143 Bondy 
Fixe : 0148027901
Mobile: 0688334906

**Comment by AlexisDrogoul** [comment]
@YoucefSklab @gnoubi @benoitgaudou could you update the status of this issue ? Is it still happening ? Are there developments that could allow for a quick fix ?

**Comment by benoitgaudou** [comment]
I just tested on yesterday's GAMA 1.8.2 Alpha release and it is still occurring.

**Comment by lesquoyb** [comment]
So far here is where I am with this one on the git version of gama : 

- as mentioned by @benoitgaudou it appears only when we send an agent with a non-empty mailbox to another agent (an agent sending a message to himself seems to not go through the same process, also everything works well sending an agent with an empty mailbox)
- the serialization part works well, it fails at deserializing
- joining the xstream source jar to the code I have been able to investigate further: it actually fails when trying to build the sender of a message in the mailbox, which refers to the agent in whom the message is stored

Here is a minimal test model to reproduce the bug:
sender:
```

model sender


global skills:[network]{	

	init {
		do connect  with_name:"sender";
	}


	reflex send1 when: cycle = 0 {
		write "sending first message";
		do send to:"sender" contents:"first message";
	}
	
	reflex receive
	{
		write "mail box content"  + mailbox collect(each.contents);
	}
	
	reflex send2 when: cycle > 0 {
		write "sending itself to receiver";
		do send to:"receiver" contents:self;	
	}

}


experiment Network_sender type: gui {
	output {
	}
}
```
receiver:

```
model receiver


global skills:[network]{	
	
	init {
		do connect  with_name:"receiver";
	}
	
	reflex fetch when:has_more_message()
	{	
		message mess <- fetch_message();
		write name + " fetch this message: " + mess.contents;	
	}	
}



experiment Network_receiver type: gui {
	output {
	}
}
```
The sent serialized message looks like this:
```
<msi.gama.kernel.simulation.SimulationAgent>
  <index>0</index>
  <variables>
    ...
    <data>
      <msi.gama.metamodel.shape.GamaShape>
        ...
        <attributes class="msi.gama.util.GamaMap">
          ...
          <valuesMapReducer>
            <entry>
              <string>messaging_skill_mailbox</string>
              <msi.gama.extensions.messaging.GamaMailbox>
                <valuesListReducer>
                  <msi.gama.extensions.messaging.GamaMessage>
                    <unread>false</unread>
                    <sender class="msi.gama.kernel.simulation.SimulationAgent" reference="../../../../../../../../../.."/>
                    <receivers class="string">sender</receivers>
                    <contents class="string">first message</contents>
                    <emissionTimeStamp>0</emissionTimeStamp>
                  </msi.gama.extensions.messaging.GamaMessage>
                </valuesListReducer>
                <contentTypeListReducer class="msi.gaml.types.GamaMessageType">
                  <GamaType>message</GamaType>
                </contentTypeListReducer>
              </msi.gama.extensions.messaging.GamaMailbox>
            </entry>
          </valuesMapReducer>
        </attributes>
      </msi.gama.metamodel.shape.GamaShape>
      ...
    </data>
  </variables>
  ...
</msi.gama.kernel.simulation.SimulationAgent>
```
And the error occurs when trying to deserialize the `sender` tag: the reference path is correct and leads to the initial SimulationAgent, but for some reason it is not present in the converter's values list (actually there is a record for this reference but it has the value null). 
I managed to build a workaround, but there are 2 problems to it :
- it's not clean (it only answers this precise case, and probably not following the xstream philosophy which could cause other issues later)
- It imposes some limitations, for example you can't cast a message containing a "faulty" agent as a string or it would raise a StackOverflowException. 

We could just change the example model a bit so the error won't happen, or even just focus on a specific workaround for this model as I tried to do. But I think serialization will be an important feature in the future (for the headless for example) so finding what exactly is at fault here and solving the general problem will be required one day or another.

So I wonder if someone has some pointers on this? maybe @hqnghi88 ?

**Comment by hqnghi88** [comment]
Hi, I can try, still CC with the real experts. But why are you workaround "specific for a precise case"? what make this message deserialize different with a "general case"??

**Comment by lesquoyb** [comment]
Basically it works and doesn't break anything, but that's black magic to me. 
It works by pre-feeding the xstream class that's in charge of deserializing with some reference to an agent. 
I did it because I ran the code step by step in debug mode and understood what was lacking in this precise example, but I have no idea why it was the case as that same xstream class has no problem with other objects referencing themselves (as a matter of fact, the part of the code that is lacking the agent's reference is specifically commented as a part that checks for circular references and is a part of xstream that has not been updated since 2014 so I think it should be "safe"). 
So the problem could emerge again in a slightly different context, for example with an other object than an agent, and wouldn't be covered by this fix.
